source/classification1.md

```diff
@@ -1708,6 +1708,89 @@ glue("fig:05-upsample-plot", (rare_plot + upsampled_plot))
+Unfortunately, as common as it is, handling missing data properly is very
+challenging and generally relies on expert knowledge about the data, setting,
+and how the data were collected. One typical challenge with missing data is
+that missing entries can be *informative*: the very fact that an entries were
+missing is related to the values of other variables.  For example, survey
+participants from a marginalized group of people may be less likely to respond
+to certain kinds of questions if they fear that answering honestly will come
+with negative consequences. In that case, if we were to simply throw away data
+with missing entries, we would bias the conclusions of the survey by
+inadvertently removing many members of that group of respondents.  So ignoring
+this issue in real problems can easily lead to misleading analyses, with
+detrimental impacts.  In this book, we will cover only those techniques for
+dealing with missing entries in situations where missing entries are just
+"randomly missing", i.e., where the fact that certain entries are missing
+*isn't related to anything else* about the observation.
+
+Let's load and examine a modified subset of the tumor image data
+that has a few missing entries:
+
+```{code-cell} ipython3
+missing_cancer = pd.read_csv("data/wdbc_missing.csv")[['Class', 'Radius', 'Texture', 'Perimeter']]
+missing_cancer['Class'] = missing_cancer['Class'].replace({
+   'M' : 'Malignant',
+   'B' : 'Benign'
+}).astype('category')
+missing_cancer
+```
+
+Recall that K-nearest neighbor classification makes predictions by computing
+the straight-line distance to nearby training observations, and hence requires
+access to the values of *all* variables for *all* observations in the training
+data.  So how can we perform K-nearest neighbor classification in the presence
+of missing data?  Well, since there are not too many observations with missing
+entries, one option is to simply remove those observations prior to building
+the K-nearest neighbor classifier. We can accomplish this by using the
+`dropna` method prior to working with the data.
+
+```{code-cell} ipython3
+no_missing_cancer = missing_cancer.dropna()
+no_missing_cancer
+```
+
+However, this strategy will not work when many of the rows have missing
+entries, as we may end up throwing away too much data. In this case, another
+possible approach is to *impute* the missing entries, i.e., fill in synthetic
+values based on the other observations in the data set. One reasonable choice
+is to perform *mean imputation*, where missing entries are filled in using the
+mean of the present entries in each variable. To perform mean imputation, we
+use a `SimpleImputer` transformer with the default arguments, and wrap it in a 
+`ColumnTransformer` to indicate which columns need imputation.
+
+```{code-cell} ipython3
+from sklearn.impute import SimpleImputer
+
+preprocessor = make_column_transformer(
+    (SimpleImputer(), ["Radius", "Texture", "Perimeter"]),
```

Comment:
```suggestion
    (SimpleImputer(), ["Radius", "Texture", "Perimeter"]),
    verbose_feature_names_out=False
```

Currently the output table has a long prefix for the column names:

![image](https://github.com/UBC-DSCI/introduction-to-datascience-python/assets/4560057/25daabf1-2eba-44ee-9764-5e18a0c66b81)

Setting `verbose_feature_names_out=False` removes the prefix as discussed in https://github.com/UBC-DSCI/introduction-to-datascience-python/issues/146. It is a bit inconvenient to set it each time we use a column transformer, but having the long column prefix is also a bit inconvenient and makes tables take up a lot of space. I guess we could argue that there is a small benefit in clarity when each column is prefixed with their preprocessing step, but it seems mostly like a nuisance to me... Although I am a bit unsure here, what do you think?

Reply:
Sorry I didn't see your comment before posting my second one. Let's go with the shorter names.